doc: add topic - event loop, timers, `nextTick()`

techjeffharris opened the pull request:
This is a ~~_rough_~~ draft of a [Topic](https://github.com/nodejs/docs/blob/master/GETTING-STARTED.md#what-we-write) that provides an
overview of the event loop, timers, and `process.nextTick()` that is
based upon a NodeSource "Need to Node" presentation hosted by
@trevnorris: [Event Scheduling and the Node.js Event Loop](https://nodesource.com/resources).

I'm sure there's room for improvement, but this is a start!

PTAL

/cc @nodejs/documentation, @nodejs/collaborators

Qard commented:
LGTM :+1:

chrisdickinson commented:
Reviewing now — at first glance this looks like great info!

chrisdickinson commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+# Overview of the Event Loop, Timers, and `process.nextTick()`
```
This file should end with a `.md` extension.

techjeffharris replied:
ooops >_<

chrisdickinson commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+*There is a slight discrepancy between the Windows and the Unix/Linux
+implementation, but that's not important for this demonstration.  The most
+important parts are here.  There are actually seven or eight steps, but the
+ones we care about--ones that Node actually uses are these four.*
```
Is this "what Node uses vs. what libuv provides?"

techjeffharris replied:
@trevnorris will you clarify this, please?

Fishrock123 replied:
I don't think so. There are other minor steps inbetween these four provided, but as I understand, they are not particularly notable to an end user. (i.e. how this interacts with the microtask queue.)

It is possible this should be expanded to encompass some of those, so long as it is still concise.

trevnorris replied:
libuv's event loop does have more steps than what is provided here. For example, `uv_handle_t`'s are not cleaned up until the (near) final phase of the event loop `uv__run_closing_handles()`. node has no hooks into this phase, so I don't deem it note worthy.

Or that if the event loop is run with `UV_RUN_ONCE` (which node does) then there's an additional step of updating and running timers. This is also not notable because the first step of the event loop after looping around is to also process timers.

trevnorris replied:
Also there's `uv__run_idle()` phase, but is only used in node to make sure `uv__io_poll()` doesn't block in case a `setImmediate()` has been schedules. Is just implementation details that would only confuse the basic understanding of how the event loop works.

chrisdickinson replied:
> I don't think so. There are other minor steps inbetween these four provided, but as I understand, they are not particularly notable to an end user. (i.e. how this interacts with the microtask queue.)

I'd lean towards documenting microtask queue behavior if it's feasible to do so here — I've definitely found that information useful in the past when dealing with builtin promises.

chrisdickinson replied:
(But otherwise I'm totally on board with the simplification. We should note in the doc that libuv is providing the seven or eight steps, but that Node only exposes the steps described in this document to end users, primarily.)

trevnorris replied:
> I'd lean towards documenting microtask queue behavior if it's feasible to do so here — I've definitely found that information useful in the past when dealing with builtin promises.

I think the most notable thing is that it's processed after the nextTick queue, but then after the microtask queue is processed the nextTick queue is processed again. Then back to the microtask queue. etc.

This is to handle the case that one queues the other. But might be sufficient to put that it's "processed at the same time as the nextTick queue".

chrisdickinson commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+
+Looking back at our diagram, any time you call nextTick in a given phase, all
+callbacks passed to nextTick will be resolved before the event loop continues.
+This can create some bad situations because **it allows you to asynchronously
```
I might suggest nixing `asynchronously` here, since it's synchronously blocking asynchronous events.

chrisdickinson commented on doc/topics/the-event-loop-timers-and-nexttick:
````diff
@@ -0,0 +1,174 @@
+}
+```
+
+The snippet does an argument check and if its not correct, it will pass the
````
`it's`

chrisdickinson commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+to be propagated as the arguments to the callback so you don't have to nest functions.
+
+What we're doing is passing an error back to the user.  As far as the event loop
+is concerned, its happening **synchronously**, but as far as the user is
```
Might consider splitting this sentence in two:

```
As far as the _event loop_ is concerned, this happens **synchronously**. However,
as far as the _user_ is concerned, it occurs **asynchronously**: `apiCall()` was written
to always run its callback *after* the rest of the user's code.
```

techjeffharris replied:
:+1: Thank you!

chrisdickinson commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+emit an event from the constructor immediately because the script will not have
+processed to the point where the user assigns a callback to that event.  So,
+within the constructor itself, you can set a callback to emit the event after
+the constructor has finished, which provides the expected results.
```
Wouldn't one want to use `setImmediate` in both of these cases?

tomgco replied:
1. Should use `setImmediate` in this case.
2. We shouldnt use nexttick for sending events, instead it should be used for setting up listeners for events so that they always exist on the next tick of the event loop, the example you provided would be a good use case for setImmediate.

Maybe also an example as well would be awesome :D

techjeffharris replied:
It sounds like `setImmediate` should pretty much always be the choice...

Are there good reasons to use `nextTick` then? Or is it basically deprecated, but not really?

trevnorris replied:
Just wrote up something about this: https://github.com/nodejs/docs/issues/34#issuecomment-176902809

trevnorris replied:
@chrisdickinson As for the constructor case, say the construction is happening in a `setImmediate()` and you don't emit until the next `setImmediate()`. That means an I/O related event could be triggered in the polling phase, thus causing events to fire in incorrect order.

techjeffharris replied:
This begs the questions:
* does the event loop start *once* the input script has **finished** execution **--OR--**
* does the event loop start *then* execute the input script, and if so, in which 'phase' is the input script executed? In the `poll` phase...?

If it is the former, then it seems to me that `nextTick()` would be the correct choice because it will be called once the input script has finished executing which prevents any possibility of I/O events being processed before your callback is executed, as described here: https://github.com/nodejs/docs/issues/34#issuecomment-176902809

If it is the latter and it IS in the `poll` phase, then it seems it wouldn't make much difference since the `setImmediate` phase is right after the `poll` phase.  

From what I've gathered, I believe the former provides the most accurate analogy for this example.

trevnorris replied:
@techjeffharris The event loop doesn't start until after bootstrap. Which includes the initial script passed to node. Here's a partial native stack of running `node -e 'process.abort'`:
```
    frame #2: 0x00000000015e00d1 node_g`node::Abort(args=0x00007fffffffc160) + 17 at node.cc:1643
...
    frame #34: 0x0000000000ccfce1 node_g`v8::Function::Call(this=0x0000000001efe550, recv=(val_ = 0x0000000001efe330), argc=0, argv=0x0000000000000000) + 113 at api.cc:4422
    frame #35: 0x00000000015c7732 node_g`node::Environment::KickNextTick(this=0x0000000001ef2ae0) + 242 at env.cc:81
    frame #36: 0x00000000015d9f1c node_g`node::MakeCallback(env=0x0000000001ef2ae0, recv=(val_ = 0x0000000001efe330), callback=(val_ = 0x0000000001ef0ae8), argc=2, argv=0x00007fffffffd9d0) + 1916 at node.cc:1192
    frame #37: 0x00000000015da166 node_g`node::MakeCallback(env=0x0000000001ef2ae0, recv=(val_ = 0x0000000001efe330), symbol=(val_ = 0x0000000001ef0ad8), argc=2, argv=0x00007fffffffd9d0) + 214 at node.cc:1218
    frame #38: 0x00000000015da1e1 node_g`node::MakeCallback(env=0x0000000001ef2ae0, recv=(val_ = 0x0000000001efe330), method=0x00000000017f7aa6, argc=2, argv=0x00007fffffffd9d0) + 97 at node.cc:1228
    frame #39: 0x00000000015e359d node_g`node::EmitBeforeExit(env=0x0000000001ef2ae0) + 349 at node.cc:3893
    frame #40: 0x00000000015e40af node_g`node::StartNodeInstance(arg=0x00007fffffffdbb0) + 719 at node.cc:4088
    frame #41: 0x00000000015e3d56 node_g`node::Start(argc=1, argv=0x0000000001ea9ad0) + 262 at node.cc:4160
    frame #42: 0x0000000001618f72 node_g`main(argc=3, argv=0x00007fffffffdd18) + 34 at node_main.cc:44
```
If it had been in the event loop then you would have seen `uv_run()`.

Even if you're in the poll phase then execution order of anything queued while _in_ a `setImmediate` callback will be different from anything outside of it. It's always relevant to use `nextTick` in these cases.

techjeffharris replied:
Excellent, thank you for clarifying that.

chrisdickinson commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+potential breakages occur.  While they are confusing, the names themselves won't change.
+
+*We recommend developers use setImmediate in all cases because its easier to
+reason about.*
```
(And it leads to code that's compatible with a wider variety of environments, like browser JS.)

trevnorris replied:
technically this can be achieved using `Promise.resolve(1).then(...)`, which runs in similar fashion to the `nextTickQueue`.

chrisdickinson commented:
This is excellent work!

mscdex commented:
It might be worth mentioning that `setImmediate()` was introduced in node v0.10? Whereas `process.nextTick()` has been around for much longer.

techjeffharris commented:
Thanks for the notes!  I must go to my day (evening) job ATM, but I'll address all concerns mentioned ASAP.

Fishrock123 commented:
@trevnorris should probably review this also as he has previously done detailed analysis of the event loop for presentations and such.

shigeki commented:
I made a diagram in several years ago for the discussion in https://html5.ohtsu.org/node-v0.11.13-eventloop.png
I hope it can be helpful for understanding for an event loop in libuv.

jasnell commented:
+1 on this... awesome to see.

ronkorving commented:
Fantastic work :+1:

tomgco commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+
+### Why would that be allowed?
+
+Why would something like this be included in Node?  Part of it is a design
```
Node.js

tomgco commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+
+// the callback is called before `someAsyncApiCall` completes.
+someAsyncApiCall(function () {
+
```
Maybe we should use ES2015 for our examples? IE fat arrow functions?

Fishrock123 commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+    └──│     setImmediate      │
+       └───────────────────────┘
+
+note: each box will be referred to as a "phase" of the event loop.
```
Maybe make this italic?

techjeffharris commented:
Pushed some updates to address issues/concerns/suggestions provided above.

calvinmetcalf commented:
you need to remove the file without the .md

techjeffharris commented:
> I'm happy to answer questions on node.js / libuv / V8 for anyone doing a write-up.

>_<  thanks!  Still a bit of a git novice--learning :D

techjeffharris commented:
Also, sorry for the zealous rebases for a simple doc PR....

trevnorris commented on doc/topics/the-event-loop-timers-and-nexttick:
```diff
@@ -0,0 +1,174 @@
+eventually enter the 'poll' phase which determines how many milliseconds remain
+until the next timer. If there is a timer, it will wait for connections for that
+many milliseconds. After that many milliseconds, it will break the 'poll' phase
+and wrap back around to the timers phase where those callbacks can be processed.
```
When the poll phase of the event loop is entered the number of ms before the soonest timer is to be called is set as the poll's timeout. Meaning the poll phase will return after "timeout" ms.

Take note that the poll phase can only return while idle. Meaning execution of a callback is allowed to run to completion, and can cause unexpected delay running the timer.

techjeffharris replied:
:+1: Updated to include these notes.  Thanks for the further detail!

trevnorris commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+However, after a callback has been scheduled with `setImmediate()`, at the
+start of the poll phase, a check will be run to see if there are any callbacks
+waiting.  If there are none waiting, the poll phase will end and continue to
+the `setImmediate` callback phase.
```
Just want to make sure it's clear enough to readers that if any I/O has been queued prior to arriving at the poll phase, those will be allowed to run. The poll phase isn't skipped entirely because a `setImmediate()` has been scheduled.

bnoordhuis replied:
IOW, scheduling one or more `setImmediate()` callbacks mean libuv still polls for I/O but it won't sleep in the poll system call.

techjeffharris replied:
Duly noted.  

At the timers and setImmediate phases, each respective queue is completely exhausted before the phase completes; however, it appears that the poll phase executes only one callback per iteration, correct?

for example: 

```js
// pretend these are the queues used by the event loop and assume that none of these 
// timers/callbacks/immediates schedule any _more_ timers/callbacks/immediates...

let timers_queue = [ timer1, timer2, timer3];
let poll_queue = [callback1, callback2, callback3];
let immediate_queue = [immediate1, immediate2, immediate3];

// also ignoring the 'pending callbacks' phase...
// iteration 1: 
// [timers phase begins]
timer1();
timer2();
timer3();

// [poll phase begins]
// callback1();

// [immediate phase beings]
// immediate1();
// immediate2();
// immediate3();

// iteration 2: 
// [timers phase begins]
// no timers queued

// [poll phase begins]
// callback2();

// [immediate phase beings]
// no immediates queued

// merry-go-round continues..

trevnorris replied:
There are a set number of poll events that are allowed to be processed before execution is forced to continue. On Linux I believe it's `48 * 1024` (https://github.com/libuv/libuv/blob/v1.8.0/src/unix/linux-core.c#L220) /cc @bnoordhuis got that from `events[1024]` times the maximum number of loops that is allowed to take.

bnoordhuis replied:
Not quite right - there can be many I/O callbacks in the poll phase.

bnoordhuis commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+milliseconds, the `poll` phase will return and wrap back around to the 
+timers phase where those callbacks can be processed.
+
+Take note that the poll phase can only return while idle; execution of 
```
'can only return' is kind of confusing, IMO.  I think you mean to say that node.js can only enter the poll phase when otherwise idle?  (Which is mostly correct, although libuv can and will poll for I/O without blocking, i.e., poll with a zero millisecond timeout.)

trevnorris commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+
+This is the phase in which the event loop sits and waits for incoming
+connections to be received.  Ideally, most scripts spend most of their time
+here.
```
May want to make note of the fact that new events processed in poll are queued by the kernel. So while polling events are being processed, more poll events can be queued. So long running callbacks can allow the poll phase to run even longer. Though libuv also has a hard maximum (system dependent) before it stops poll'ing for more events. This is so the poll phase cannot starve the event loop.

Maybe that's too much information for this doc, but figured I'd mention it.

bnoordhuis commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+delay running the timer. 
+
+*Note: The `poll` phase technically controls when timers are called due to its
+ability to cause a thread to sit idly without burning CPU in order to stall the
```
'to sleep'?  Also, 'stall' doesn't seem like the right word because it's the opposite; when the event loop wakes up is when timers can run.

bnoordhuis commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+## poll:
+
+This is the phase in which the event loop sits and waits for incoming
+connections to be received.  Ideally, most scripts spend most of their time
```
'received'.  It's not just connections - completion of async file I/O, DNS operations, etc. cause libuv to wake up from the poll system call.

bnoordhuis commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+
+How quickly a `setImmediate()` callback is executed is only limited by how
+quickly the event loop can be processed whereas a timer won't fire until the
+number of milliseconds passed have elapsed.
```
As a bit of background, it would be more technically correct to say 'at least the number of milliseconds have passed' because timers can fire late due to process/thread scheduling issues.

trevnorris commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+What we're doing is passing an error back to the user.  As far as the _event
+loop_ is concerned, this happens **synchronously**. However, as far as the
+_user_ is concerned, it occurs **asynchronously**: `apiCall()` always runs its
+callback *after* the rest of the user's code.
```
Possibly add that this is achieved by allowing the JS call stack to unwind, but then calling another JS function immediately afterward. This is why you can make recursive calls to `nextTick` without reaching a `RangeError: Maximum call stack size exceeded` from v8.

trevnorris commented on doc/topics/the-event-loop-timers-and-nexttick.md:
````diff
@@ -0,0 +1,183 @@
+});
+
+var bar = 1;
+```
````
A real world example in node would be:
```js
const server = net.createServer(() => {}).listen(8080);

server.on('listening', () => {});
```
When only a port is passed the port is bound immediately. So the `'listening'` callback could be called immediately. Problem is that the `.on('listening'` will not have been set by that time.  
To get around this the `'listening'` event is queued in a `nextTick()` to allow the script to run to completion. Which allows the user to set any event handlers they want.

techjeffharris replied:
I added your real world example in https://github.com/techjeffharris/node/commit/f80d7cc7fe12ffada96c4c2d9fbfb3d1b47b36f1

trevnorris commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+By placing it in a `process.nextTick()`, the script still has the ability to
+run to completion, allowing all the variables, functions, etc., to be
+initialized prior to the callback being called.   It also has the advantage of
+not allowing the event loop_ to continue.  It may be useful that the user be
```
is it supposed to be `_event loop_`? (missing the `_`)

techjeffharris replied:
I believe its a typo as I cannot see a reason to italicize "event loop" in this case.

bnoordhuis commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+You may have noticed that `process.nextTick()` was not displayed in the
+diagram, even though its a part of the asynchronous API.  This is because
+`process.nextTick()` is not technically part of the event loop.  Instead, it
+is executed at the end of each phase of the event loop.
```
It's actually more complex than that: nextTick callbacks are dispatched after every call into the VM through `node::MakeCallback()`.  Calls are made for lots of things; incoming connections, incoming data, after write callbacks, (some) timers, etc.

trevnorris replied:
@bnoordhuis an aside: i'm working on having `MakeCallback` not process the `nextTickQueue` every time in an effort to make it reentrant.

@techjeffharris There's also the case that if there's a handled exception while processing the `nextTickQueue` then execution will be deferred until later. Meaning the queue will not have been completely drained.

techjeffharris replied:
@bnoordhuis Would it be more appropriate to say something like "the nextTickQueue will be processed after the current callback executes, regardless of the current event loop phase."  ...?

bnoordhuis replied:
Yes, that's fairly correct, although 'callback' here doesn't necessarily mean 'user callback'; it might be some bookkeeping the node runtime is doing.

bnoordhuis commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+don't have to nest functions.
+
+What we're doing is passing an error back to the user.  As far as the _event
+loop_ is concerned, this happens **synchronously**. However, as far as the
```
I'm not sure I understand why it would be synchronous from the event loop's perspective.  There's a discontinuity so I'd label it asynchronous.

trevnorris replied:
I think he means that queuing callbacks with `nextTick` will not allow the event loop to proceed from say `uv__io_poll()` to `uv__run_check()`. So from the `uv_run()` perspective it would be synchronous. Have a better way of describing that?

techjeffharris commented:
Gosh, I really have my work cut out for me!  I thin I'm going to let the dust settle before I traverse these comments/suggestions and make adjustments.. :)

Thank you all for the excellent feedback!

trevnorris commented:
@techjeffharris It's great that you've tackled this. It will be nice to have such a technically correct document describing how the event loop works.

techjeffharris commented:
>  It's great that you've tackled this.

There's still a lot of work to be done before I can feel like I have really "tacked" this document.  ;)

Nonetheless, I'm happy to have a way to help the community while improving my understanding of such fundamental features as these!

techjeffharris commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,183 @@
+
+## pending callbacks:
+
+This phase executes callbacks for specific types of TCP errors, for example.
```
@bnoordhuis, @trevnorris, could either of you provide me some good examples of operations that will execute in the `pending callbacks` phase, please?  This section is sadly lacking..

trevnorris replied:
For example if tcp connect receives a `ECONNREFUSED`, some *nix systems want to wait to report the error. This will be queued to execute in the "pending" phase.

@bnoordhuis You may have better examples. I'm just going from libuv source to pull these out. But haven't been able to write a script that demonstrates this precisely.

techjeffharris commented:
It appears that more detail is needed to describe the basic progression through each phase of the event loop before I describe how each phase works.  This will probably manifest as something along the lines of:

"Each phase has a queue of operations executed FIFO.  Each operation in the queue must complete execution before moving to the next phase in the event loop.  Since any of these operations may schedule more operations, _exactly_ when a given callback will be executed is often **indeterminate**."  

Once a bit of background for each phase is described, I can go into examples describing some of the nuances.

Addressing comments, concerns, suggestions from @bnoordhuis and @trevnorris:

_Please note that the placement of various details refined below will probably evolve as I put together a second draft._

- [ ] https://github.com/nodejs/node/pull/4936/files#r51319963
  I think I'll reword this whole section to something like: 

  "Contrary to what a person might expect, a timer specifies the threshold (_minimum_ time rather than  _exact_ time) after which the provided callback **may** be executed. Therefore, if many long-running I/O operations are waiting in the `poll` queue, the time that elapses prior to the callback being executed could be far greater than the threshold specified. 
  
  "For example, say you schedule a timeout to execute after a 100 ms threshold.  When the event loop enters the `poll` phase, _if there are 0 I/O events queued_, the process will wait for I/O operations for the remaining (say 95) of ms of the threshold.  If after 90 of those ms, an I/O operation that takes 10ms to complete is added to the queue, that operation will run to completion.  Again, if there are 0 I/O events in the queue, the poll phase will check for timers _whose time thresholds have elapsed_ and wrap back to the timers phase to execute those timers' callbacks.  In this case, the total delay between the timer being scheduled and its callback being executed will be 105ms (in addition to the the time the event loop itself takes to orchestrate this.)"

- [ ] https://github.com/nodejs/node/pull/4936#discussion_r51318764 ... https://github.com/nodejs/node/pull/4936#discussion_r51323157

  I'll make note that when using `setImmediate()`, the delay prior to the setImmediate callback being executed depends on 
  * When `setImmediate()` is called -- during startup, a `nextTick()` call, a timer callback, etc.
  * the number of I/O operations that have been queued (0 or more), and 
  * how long those operations take to complete.  

  In essence: if you want to _guarantee_ that your desired operation is executed just after the current code, you should use `nextTick()`, and if that doesn't matter you should probably use `setImmediate()`.  I'll probably add a link titled "see examples in `timers` phase" also.


- [ ] https://github.com/nodejs/node/pull/4936/files#r51320733

  "The `poll` phase technically controls when timers are executed.  While waiting for the minimum threshold specified when scheduling the timer, the `poll` phase will iterative through its queue of callbacks executing them sequentially until either the queue is exhausted, or the system-dependent hard limit is reached.  Collectively, these callbacks _may or may not_ take longer to complete than the minimum threshold specified by the timer.  The actual amount of time depends on how many callbacks end up being executed and how long they take.  Furthermore, since new events processed in `poll` are queued by the kernel, more poll events can be queued while polling events are being processed.  As a result, long running callbacks can allow the poll phase to run much longer than a timer's threshold.  See [`poll` phase](#poll) section for more details"

  "Note: To prevent the `poll` phase from starving the event loop, libuv also has a hard maximum (system dependent) before it stops poll'ing for more events."

- [ ] https://github.com/nodejs/node/pull/4936/files#r51320851 
  How about something more like: 
  "This is the phase in which the event loop either: 
    * sequentially processes events in its queue, --OR--
    * sits and waits for incoming connections to be received, completion of async file I/O, DNS operations, etc."

- [ ] https://github.com/nodejs/node/pull/4936/files#r51318764 ... https://github.com/nodejs/node/pull/4936/files#r51323157

  How does this sound?

  "Generally, as the code is executed, the event loop will eventually advance to the `poll` phase where it will either process callbacks in its queue or wait for an incoming connection, request, etc. as usual.   However, if a callback has been scheduled with `setImmediate()`, at the start of the next poll phase a check will be run to see if there are any callbacks waiting in the queue and one of two things will happen: 

  * If there are no callbacks waiting in the queue, the poll phase will end and continue to `setImmediate` phase.  In other words, scheduling one or more `setImmediate()` callbacks tells libuv to `poll` for I/O events but prevents it from "sleeping" to wait for I/O events/timers so that the `setImmediate()` callback may be executed, or
  * If any I/O has been queued prior to arriving at the poll phase, those will be allowed to run meaning the poll phase isn't skipped entirely because a setImmediate() has been scheduled."  

- [ ] https://github.com/nodejs/node/pull/4936/files#r51321347 
  
  Addressed above ^^

- [ ] https://github.com/nodejs/node/pull/4936/files#r51322215 ... https://github.com/nodejs/node/pull/4936/files#r51323639

  Need to improve my understanding a bit more before I can improve this section. 

- [ ] https://github.com/nodejs/node/pull/4936/files#r51322409 ... https://github.com/nodejs/node/pull/4936/files#r51321529

  I think I'll reword this since its hard to make accurate generalizations about the event loop's "perspective".  How does something like this sound: 

  "What we're doing is passing an error back to the user but only *after* we have allowed the rest of the user's code to execute.  By using `process.nextTick()` we guarantee that `apiCall()` always runs its callback *after* the rest of the user's code and *before* the event loop is allowed to proceed.  This is achieved by allowing the JS call stack to unwind, but then executing the provided callback immediately afterward which allows a person to make recursive calls to nextTick without reaching a `RangeError: Maximum call stack size exceeded from v8`."
  
  Aside: This makes me think that maybe this Topic should be catered toward the Node.js novice and that we should have a detailed walkthrough geared for more advanced users that goes much further into detail than this overview document--for example: _all_ steps in the event loop, exactly how many callbacks may be executed in the poll phase, the various queues, etc.

techjeffharris commented:
@mscdex 

> It might be worth mentioning that setImmediate() was introduced in node v0.10? Whereas process.nextTick() has been around for much longer.

I will find a good place to note this!

techjeffharris commented:
@shigeki Thank you!  This is very helpful!

mcollina commented:
Excellent work.

bnoordhuis commented:
> Aside: This makes me think that maybe this Topic should be catered toward the Node.js novice and that we should have a detailed walkthrough geared for more advanced users that goes much further into detail than this overview document--for example: all steps in the event loop, exactly how many callbacks may be executed in the poll phase, the various queues, etc.

Great idea!

techjeffharris commented:
Just pushed an update that hopefully addresses all comments/concerns above.  PTAL.

Sorry this has been such a long process!  

@nodejs/documentation @bnoordhuis @trevnorris

davepacheco commented:
Thanks for doing this! There's a lot of really good information here.  I think it's great to elevate the event loop semantics to a top-level topic.  Here are a few thoughts:

Where you use "the kernel", I think we may want to say "the system", or something more general like that.  The asynchronous operation may be implemented by the kernel, or it may be implemented by an add-on that makes no use of kernel facilities at all.

I wonder if we could describe the interface in a way that avoids users having to know so much arcane implementation detail.  For example, the doc talks about phases that I'm not sure a Node user needs to distinguish between, and then it explains that some of these phases are basically skipped depending on the state of other phases.  (Specifically, there's the poll phase, followed by the setImmediate phase, but the second paragraph in the `setImmediate` section explains that the poll phase is essentially skipped if there are immediate callbacks pending.)  I assume the existing description does reflect the implementation, but I wonder if we couldn't simplify the presentation without losing any useful information for users.

Specifically, is there a reason that the timers phase is presented first (other than the fact that the implementation may be structured that way)?  Is it accurate to say that callbacks are processed in this order: immediate callbacks, timer callbacks, "pending callbacks", then callbacks associated with asynchronous operations ("poll callbacks"), and that when there are no more callbacks pending, Node waits for asynchronous operations to complete?

I think the summary of setImmediate() and nextTick() is great.  Another thing you might want to mention is that if you choose to invoke a user's callback with `nextTick()` rather than `setImmediate()`, and the user callback calls back into you and you use nextTick() again, that can easily go into a loop that starves the event loop.  The reason I think this case is important is because neither you nor the callback author could reasonably have known that this would happen.  The nextTick() user is really the only one in a position to know this _could_ happen, so this is another reason I avoid using nextTick() for callbacks that I don't know anything about.

I hope this is helpful.  Thanks again for writing the docs.

a0viedo commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,358 @@
+value you may set a timer's delay to is 1 ms (0 is coerced to 1) which doesn't
+seem like much time to us humans, but it's actually pretty slow compared to
+how quickly `setImmediate()` can execute — the event loop operates on the
+microsecond scale (1 ms = 1000 µs.)
```
Given that it cannot be assured an exact time that `setImmediate` will run before `setTimeout` I think this could be simplified by saying that the main advantage would be that `setImmediate` has _high probability_ of running  before `setTimeout`.

techjeffharris replied:
Unfortunately, it would appear that neither of us are quite right..

I just created an example to demonstrate the actual activity.  It turns out that the order in which the callbacks will execute is non-deterministic when they are scheduled in the same function at startup.  There may be a slight preference toward whichever is called first, but I did not gather or analyze any stats, so it could just be my brain recognizing non-existent patterns. :)

time_vs_immediate.js:
```js
setTimeout(function timeout () {
  console.log('timeout');
},0);

setImmediate(function immediate () {
  console.log('immediate');
});
```

    $ node timeout_vs_immediate.js
    timeout
    immediate

    $ node timeout_vs_immediate.js
    immediate
    timeout
    

However, when I add a 1ms delay after `setTimeout` and `setImmediate` have been called, the timeout callback will always be executed first: 

time_vs_immediate.js:
```js
setTimeout(function timeout () {
  console.log('timeout');
},0);

setImmediate(function immediate () {
  console.log('immediate');
});

const start = Date.now();

// allow >= 1ms elapse
while (start === Date.now()) {
  ; // no-op
}
```

    $ node timeout_vs_immediate.js
    timeout
    immediate

    $ node timeout_vs_immediate.js
    timeout
    immediate

mcollina replied:
The `setImmediate` list is executed once per loop iteration, right after polling for i/o (source https://github.com/nodejs/node/blob/master/src/node.cc#L2724-L2747 and http://docs.libuv.org/en/v1.x/check.html).

There are in fact two "timers" phases in libuv https://github.com/libuv/libuv/blob/c0fa2e7518a3a0e364c56f8223bdd0f549ddac66/src/unix/core.c#L331-L380, because node runs `uv_run` with `UV_RUN_ONCE`:
https://github.com/nodejs/node/blob/master/src/node.cc#L4170.

Your examples are not happening within a i/o cycle, so you are bound to _how fast your process go_, which is impacted by other programs running on your machine. 
If you move the two calls within a i/o cycle, the immediate callback is always executed first:

```js
var fs = require('fs')

fs.readFile(__filename, () => {
  setTimeout(() => {
    console.log('timeout')
  }, 0)
  setImmediate(() => {
    console.log('immediate')
  })
})
```

@saghul @cjihrig for confirmation.

a0viedo replied:
This is discussed at https://github.com/nodejs/node-v0.x-archive/issues/6034#issuecomment-44665472 and https://github.com/nodejs/node-v0.x-archive/issues/25788

techjeffharris commented:
> if you choose to invoke a user's callback with nextTick() rather than setImmediate(), and the user callback calls back into you and you use nextTick() again, that can easily go into a loop that starves the event loop.

@davepacheco Could you give me a simple code example, please?  Here's the closest idea I have to what you're describing, but I think I'm missing something..

my_mod.js
```js 
module.exports = function (callback) {

  process.nextTick(callback);

};
```

user_script.js
```js
var mod = require('./my_mod');

function finished() {
  mod(finished);
}

mod(finished);
```

techjeffharris commented:
> Where you use "the kernel", I think we may want to say "the system", or something more general like that. The asynchronous operation may be implemented by the kernel, or it may be implemented by an add-on that makes no use of kernel facilities at all.

Maybe OS would be most appropriate.

techjeffharris commented:
> I wonder if we could describe the interface in a way that avoids users having to know so much arcane implementation detail...I assume the existing description does reflect the implementation, but I wonder if we couldn't simplify the presentation without losing any useful information for users.

The presentation order could probably be adjusted to ease comprehension of the overall process.  Indeed, the timers phase describes many nuances of the poll phase...  Maybe it would be best to describe the poll phase first, then setImmediate, etc.?

techjeffharris commented:
> Specifically, is there a reason that the timers phase is presented first (other than the fact that the implementation may be structured that way)? 

Not really, no.  I start with timers because that's where the loop starts for the purposes of this Topic. 

> Is it accurate to say that callbacks are processed in this order: immediate callbacks, timer callbacks, "pending callbacks", then callbacks associated with asynchronous operations ("poll callbacks"), and that when there are no more callbacks pending, Node waits for asynchronous operations to complete?

Simply put, yes but only after the `timers`, `pending_callbacks`, and `poll` phases have run through at least once--see https://github.com/nodejs/node/pull/4936#issuecomment-176504434.  Of course each phase has its nuances, like the poll phase will be skipped if its queue is empty and there is an immediate callback queued.

techjeffharris commented:
Looking for LGTMs from @trevnorris, @bnoordhuis, @nodejs/collaborators, @nodejs/documentation. 

Any and all feedback is appreciated!

trevnorris commented:
LGTM

mscdex commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+
+### `setImmediate()`:
+
+`setImmediate()` is actually a special timer that runs in a separate phase of
```
I'm not sure that I would call it a timer really since it has no "timeout" value per se ...

techjeffharris replied:
I just used @trevnorris' words from the presentation link in the OP, but I am open so suggestions as per his approval.

trevnorris replied:
It's located in https://nodejs.org/api/timers.html :)

mscdex commented on doc/topics/the-event-loop-timers-and-nexttick.md:
````diff
@@ -0,0 +1,375 @@
+```
+
+Say that listen() is run at the beginning of the event loop, but the 
+listening callback is placed in a setImmediate. Now, unless a hostname 
````
All instances of `setImmediate` should probably be styled like:

<pre>
`setImmediate()`
</pre>

mscdex commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+the script to run to completion. Which allows the user to set any event
+handlers they want.
+
+## process.nextTick() vs `setImmediate()`
```
All `process.nextTick()` references should be surrounded by backticks consistently.

davepacheco commented:
> Maybe OS would be most appropriate.

"OS" is inaccurate, too.  As I mentioned, it could be an add-on.

> > Is it accurate to say that callbacks are processed in this order: immediate callbacks, timer callbacks, "pending callbacks", then callbacks associated with asynchronous operations ("poll callbacks"), and that when there are no more callbacks pending, Node waits for asynchronous operations to complete?

> Simply put, yes but only after the timers, pending_callbacks, and poll phases have run through at least once--see #4936 (comment). Of course each phase has its nuances, like the poll phase will be skipped if its queue is empty and there is an immediate callback queued.

In that case, I think the summary I suggested, even with qualifications added for the first time through the loop, would make a significantly simpler presentation for end users without sacrificing accuracy.

> @davepacheco Could you give me a simple code example, please?

I'll look for a more concrete one.  I know it exists, because that's how I learned about this problem.

Generally, it would be more like:

```javascript
/* mymod.js */
exports.myfunc = function (easy_case, callback) {
	if (easy_case) {
		process.nextTick(callback);
	} else {
		/* do something expensive */
		setTimeout(100, callback);
	}
}
```

```javascript
/*
 * example: do 1000 things with concurrency 1
 */

var mymod = require('./mymod');
var ntodo = 10000;

function doOne() {
	mymod.myfunc(true, function () {
		if (--ntodo > 0) {
			doOne();
		}
	});
}

doOne();
```

It looks like on 0.12 and later, this does not starve the event loop.  On 0.10, it explodes with a RangeError.

davepacheco commented:
@techjeffharris Ironically, the concrete example I was thinking of was the one that led me to file my issue on updating the nextTick documentation.  I described it here:
https://github.com/nodejs/node-v0.x-archive/issues/6718#issue-24380711

Basically, it was a queue abstraction somebody was using with synchronous function (i.e., it invoked the callback inside the worker function, which is usually asynchronous).  The queue used process.nextTick() to dispatch the next task.  So completing a task would dispatch the next task synchronously, and that could end up starving the loop if you've got a lot of tasks queued.

silverwind commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+which may make async API calls, schedule timers, or call `process.nextTick()`, 
+then begins processing the event loop.  
+
+The Following diagram shows a simplified overview of the event loop's
```
lowercase `following`

silverwind commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+
+*note: each box will be referred to as a "phase" of the event loop.*
+
+Each phase has a FIFO queue of callbacks to execute. While each phase  is
```
double space after `phase`

silverwind commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+it will perform any operations specific to that phase,  then execute callbacks
+in that phase's queue until the queue has been exhausted or the maximum number
+of callbacks have executed.  When the queue has been exhausted or the callback
+limit is reached, the event loop will move to the next  phase, and so on.
```
Looks like there are many of these mid-sentence double spaces, please clean them up :wink:

techjeffharris replied:
Hmmm.  I suspect that's due to my auto-wrap settings for sublime text..

Thank you for bringing this to my attention!

silverwind commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+
+This phase executes callbacks for some system operations such as types of TCP
+errors.  For example if a TCP socket receives `ECONNREFUSED` when
+attempting to connect, some *nix systems want to wait to report the
```
Might need to escape the `*` to not confuse markdown parsers.

techjeffharris replied:
indeed, thank you.

yunong commented:
This is great work. Would it be better served to break this document up into several different docs? There seems to be quite a few topics covered here, and I think it would be easier to find, read, and digest were they presented separately.

Perhaps breaking out each of the sections

- Event Loop Internals
- Timers
- Callbacks
- nextTick/setImmediate API docs

For example, The current API docs around nextTick and setImmediate are very clear. We could update them with some of the content captured in this document. I know there were other PRs (https://github.com/nodejs/docs/issues/34) and (https://github.com/nodejs/node-v0.x-archive/pull/6725) that were trying to address that specifically. 

For some engineers, they may only care about the semantics of `setImmediate` and `nexttick` without wanting to fully understand the internals of the event loop immediately -- instead choose to read up on the event loop at a later time. It may be helpful to link to the event loop doc from the API docs of `setImmediate`.

Fishrock123 commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+    │  │         poll          │<─────│  connections, │
+    │  └──────────┬────────────┘      │   data, etc.  │
+    │  ┌──────────┴────────────┐      └───────────────┘
+    └──│     setImmediate      │
```
Maybe use `┤` for where the arrows originate?

trevnorris replied:
Here's my original diagram:
```
        ┌───────────────────────┐
    ╭──►│         timers        │
    │   └───────────┬───────────┘
    │   ┌───────────┴───────────┐
    │   │   pending callbacks   │
    │   └───────────┬───────────┘   ┌──────────────┐
    │   ┌───────────┴───────────┐   │  incoming:   │
    │   │          poll         │◄──┤ connections, │
    │   └───────────┬───────────┘   │  data, etc.  │
    │   ┌───────────┴───────────┐   └──────────────┘
    ╰───┤      setImmediate     │
        └───────────────────────┘
```

Fishrock123 replied:
@trevnorris ... what arrows are those?

trevnorris replied:
May want to supplement with the following:
```
        ┌───────────────────────┐
        │       bootstrap       │
        └───────────┬───────────┘
        ┌───────────┴───────────┐
    ╭──►│         timers        │
    │   └───────────┬───────────┘
    │   ┌───────────┴───────────┐
    │   │   pending callbacks   │
    │   └───────────┬───────────┘   ┌──────────────┐
    │   ┌───────────┴───────────┐   │  incoming:   │
    │   │          poll         │◄──┤ connections, │
    │   └───────────┬───────────┘   │  data, etc.  │
    │   ┌───────────┴───────────┐   └──────────────┘
    ╰───┤      setImmediate     │
        └───────────────────────┘
```
The bootstrap phase is actually important to take note of b/c there are some edge case timing changes that occur during startup.

trevnorris commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+* `pending callbacks`: this phase executes callbacks for specific types of TCP
+  errors, for example.
+* `poll`: this is the phase in which the event loop either processes
+  its queue of callbacks, or sits and waits for new callbacks for incoming
```
poll phase doesn't do anything initially on its own. (cc @bnoordhuis) It will sit and wait to be notified on a file descriptor. At which time libuv will call the corresponding callback.

trevnorris commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,375 @@
+  to the queue. Ideally, most scripts spend most of their time here. If not, 
+  it might be time consider refactoring or scaling to more processes.
+* `setImmediate`: This phase allows a person to execute callbacks
+  immediately after the `poll` phase has completed. 
```
Possibly make a note that when a setImmediate is schedules, the poll phase will still run and check for any waiting notifications. Which will then be processed. If not then execution will proceed immediately to calling the setImmediate.

sam-github commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,373 @@
+The following diagram shows a simplified overview of the event loop's
+order of operations.
+
+       ┌───────────────────────┐
```
/cc @piscisaureus Does this model adequately reflect reality?

piscisaureus replied:
The order in which certain types of callbacks are invoked is mostly determined by libuv, see https://github.com/libuv/libuv/blob/b015b4d0ee03b825d61b0b54e36aa405a8e1cf03/src/unix/core.c#L340-L366.

My take would be:

```
1. timers
2. i/o callbacks (pretty much every type of callback except timers, setImmedate, close)
2. idle, prepare (only used internally)
2. poll (= retrieve new i/o events; node will block here when appropriate)
3. check (setImmediate callbacks are invoked here)
4. close callbacks (e.g `socket.on('close', ...)`)
5. timers (again, indeed)
```

process.nextTick and microtasks are processed after every invocation of a javascript function, so they essentially run as a "loop within a loop" with a higher priority.

mcollina replied:
There is also some little activities (handling the gracefully close) done between 7 and 1 in node: https://github.com/nodejs/node/blob/master/src/node.cc#L4207-L4221.

techjeffharris replied:
@sam-github @piscisaureus @mcollina 

Again, the order of these steps is based upon this presentation by @trevnorris https://nodesource.com/resources 

Should these refinements considered show-stoppers?  I'm leaning toward getting this landed and allowing those with more intimate knowledge and understanding making some of the more specific adjustments.. Thoughts?

mcollina replied:
I consider them show-stoppers.

The order of steps is highly important, because otherwise some behaviors cannot be explained, creating more confusion.

IMHO it is acceptable to simplify things in a presentation (you have a limited time), but this should be the 'ultimate' docs on the event loop, so nobody would have to dig through all the source to understand how things works. So, let's aim for correctness first, and then prepare a "newbie friendly"  doc after (or leave it to 3rd-party blog posts).

trevnorris replied:
@piscisaureus 
> close callbacks (e.g socket.on('close', ...))

That's not always true for some odd reason. The only time the `'close'` event is emitted in `uv__run_closing_handles()` is on `Socket#_destroy()`. Otherwise it simply runs in `nextTick()` (example `Server#_emitCloseIfDrained()` and subsequent call to `emitCloseNT()`).

trevnorris replied:
@mcollina If you want more of an "ultimate" doc, here's one for reference: https://gist.github.com/trevnorris/05531d8339f8e265bd49

There are a couple updates needed (e.g. we run timers twice), but overall that represents the technical running of the event loop.

mcollina replied:
@trevnorris we are running timers twice to have more precise timings (as there might be some stuff happening between one run and the other) or is there another reason?

trevnorris replied:
@mcollina It's because we [run `uv_run()` using `UV_RUN_ONCE`](https://github.com/nodejs/node/blob/v5.9.1/src/node.cc#L4151). Which we started using [for `'beforeExit'` support](https://github.com/nodejs/node/commit/a2eeb43). The reason libuv does this is [documented in `deps/uv/src/unix/core.c`](https://github.com/nodejs/node/blob/v5.9.1/deps/uv/src/unix/core.c#L355-L365)

sam-github commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,373 @@
+
+### timers
+
+Contrary to what a person might expect, a timer specifies the **threshold**
```
node behaviour is what any software developer not on a high-end real-time OS/hardware would expect. Nothing happens at exact time in a computer, its all aproximate. You get either before or after semantics, and after is what anyone would expect. I think this text makes it sound like node is doing something odd here, when its how every commodity OS works.

benjamingr replied:
Well to be fair @sam-github, people ask a _lot_ about why their timers and timings aren't accurate. It might be worth mentioning that it's not node-specific though:

`Contrary to what a person might expect and like other server runtimes,` 

Or something like that.

sam-github replied:
Lets not quibble about the thoughfulness of peoples expectations, then :-), but just say "Timers callbacks will run as early as they can be scheduled after the specified amount of time has passed. Operating System scheduling or the running of other callbacks may delay them."

techjeffharris replied:
> Lets not quibble about the thoughfulness of peoples expectations, then :-)

Sounds good to me!  Why be presumptuous? 

> "Timers callbacks will run as early as they can be scheduled after the specified amount of time has passed. Operating System scheduling or the running of other callbacks may delay them."

I'm going to go with this ^^

sam-github commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,373 @@
+
+### `setImmediate`:
+
+`setImmediate()` is actually a special timer that runs in a separate phase of
```
its a timer? really? that surprises me, its behaviour seems unrelated to time.

benjamingr replied:
I'm not sure `timer` is good terminology for it.

techjeffharris replied:
TBH, I do not have enough intimate knowledge of the codebase to argue for or against whether or not it is a timer.  I just said its a special form of timer because that's what I got from the presentation which was the basis.  I'm hoping to get this landed ASAP, so that EVERYONE can discuss and refine this doc

mcollina replied:
`setImmediate` is not a timer, see https://github.com/nodejs/node/pull/4936#discussion_r54770391. However if we do not clarify that there is a step for executing `setImmediate` tasks, we have no way of clarifying the difference.

sam-github commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,373 @@
+## `setImmediate()` vs `setTimeout()`
+
+How quickly a `setImmediate()` callback is executed is only limited by how
+quickly the event loop can be processed whereas a timer won't fire until the
```
until "at least" the

techjeffharris replied:
fixed

evanlucas commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,373 @@
+  // use nextTick to emit the event once a handler is assigned
+  process.nextTick(() => {
+    this.emit('event');
+  }.bind(this));
```
Is the `bind` here necessary with using an arrow function?

a0viedo replied:
@evanlucas good catch. Arrow functions can't have their `this` hard-bound http://blog.getify.com/arrow-this/

techjeffharris replied:
Thanks @evanlucas !  replaced the arrow function with a plain-old anonymous function.

jasnell commented:
What's the status on this? Would love to get this landed.

techjeffharris commented:
> @jasnell: What's the status on this? Would love to get this landed.

I had been "chewing on" a few of the comments made above and how to satisfy them, but I have been working 7 days/week and so haven't had as much time to dedicate as I would prefer.

I'll put in some work today.  I would also like to get it landed so that its development is not dependent upon my availability :)

techjeffharris commented:
@jasnell - That took much less time than I was anticipating!  

PTAL - @sam-github @piscisaureus @mcollina @evanlucas @davepacheco @a0viedo @yunong @saghul @cjihrig @benjamingr 

As we are all aware, this sort of document will have a dynamic life cycle as the code base it documents changes.  That being said, I would love to get this landed so that people are able to make and suggest their own subtle edits without myself needing to be an intermediary.  

Thanks everyone for the help and support!

mcollina commented:
I am conflicted about landing this as it is. On one hand, it is a massive improvement on the current situation. On the other hand, it is incorrect in a very important part of the document.
I am also with @techjeffharris in the idea that he cannot be the gatekeeper of this PR until it is "ready", but IMHO this commit should not be part of any release before the problems above are resolved.
So if we can mark this as "don't include in a release" commit, I'm ok for landing.

benjamingr commented:
@mcollina would you be willing to fix the order of steps in this PR? If so, you can create a new PR on top of this with the fixes and we'll be able to land it.

I'm also +1 on fixing it first.

mcollina commented:
@benjamingr will do in the next few days.

techjeffharris commented:
@mcollina I have some time today and will add the steps above to overview (taking care of all the ascii art fun-stuff) and add headings to help expedite, if you would like.

mcollina commented:
Please go ahead! :)
Il giorno mar 29 mar 2016 alle 20:14 Jeff Harris <notifications@github.com>
ha scritto:

> @mcollina <https://github.com/mcollina> I have some time today and will
> add the steps above to overview (taking care of all the ascii art
> fun-stuff) and add headings to help expedite, if you would like.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/4936#issuecomment-203032848>
>

techjeffharris commented:
Added those bits I mentioned.  Ping @mcollina

mcollina commented:
@techjeffharris that was awesome! I added in some minor things: https://github.com/techjeffharris/node/pull/1.

I have just one question left for @trevnorris (see above ^^), but apart from that I think we are good.

How the event loop work is *complex*, I hope this doc will clarify the confusion in the community.

trevnorris commented:
@mcollina I answered your question above. Though I believe that level of information about our implementation details isn't useful. Especially since the two timer resolve operations happen back-to-back. So on the second pass there will always be nothing to execute (unless the `'beforeExit'` event is used).

mcollina commented:
@trevnorris so you think we can safely remove the second pass on the timer phase from this explanation?

ajafff commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,464 @@
+Unix/Linux implementation, but that's not important for this
+demonstration. The most important parts are here. There are actually
+seven or eight steps, but the ones we care about — ones that Node.js
+actually uses are these four._
```
>these four.

doesn't match your previous paragraphs. Maybe just omit this sentence.

ajafff commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,464 @@
+also has a hard maximum (system dependent) before it stops `poll`ing for
+more events.
+
+### i/o callbakcs:
```
typo in callba**ck**s
same 5 lines below

ajafff commented on doc/topics/the-event-loop-timers-and-nexttick.md:
```diff
@@ -0,0 +1,464 @@
+then the  `poll` phase becomes idle, it will end and continue to the
+`check`  phase rather than waiting for `poll` events.
+
+### `close callbacks: 
```
closing backtick missing

techjeffharris commented:
@mcollina Merged your commits
@ajafff fixed your nits that were not addressed in @mcollina's commits

mcollina commented:
LGTM if there are no other open points.

trevnorris commented:
@mcollina Removing mention of a second execution of timers has no technical impact I can see. So it should be fine to remove.

techjeffharris commented:
@trevnorris 
> @mcollina Removing mention of a second execution of timers has no technical impact I can see. So it should be fine to remove.

Good, because I already merged his change ;)

techjeffharris commented:
Okay, I believe I have exhausted my capacity to be the guardian of this PR, and it appears that all major concerns have been addressed.  Thanks for all the help!

@nodejs/collaborators, @nodejs/ctc PTAL

mcollina commented:
LGTM! :)

jasnell commented:
LGTM
If there's anything additional to be done on this it can be done as separate PRs. Awesome work, thank you!

calvinmetcalf commented:
landed

lordKnighton commented:
👍

techjeffharris commented:
Yay! Thank you all for your help!

a0viedo commented:
Excellent work @techjeffharris!

techjeffharris commented:
Thank you, @a0viedo !

TheAlphaNerd commented:
@calvinmetcalf just as a heads up this landed with a bunch of whitespace errors :cry: 

```
Applying: doc: add topic - event loop, timers, `nextTick()`
.git/rebase-apply/patch:18: trailing whitespace.
offloading operations to the system kernel whenever possible.
.git/rebase-apply/patch:26: trailing whitespace.
## Event Loop Explained
.git/rebase-apply/patch:48: trailing whitespace.
    │  ┌──────────┴────────────┐      └───────────────┘
.git/rebase-apply/patch:53: trailing whitespace.
       └───────────────────────┘
.git/rebase-apply/patch:81: trailing whitespace.
* `timers`: this phase executes callbacks scheduled by `setTimeout()`
warning: squelched 20 whitespace errors
warning: 25 lines applied after fixing whitespace errors.
```

If you are running `git am` you can avoid this by including the flag `--whitespace=fix`

silverwind commented:
Or even better, use [`apply-pr`](https://github.com/petkaantonov/apply-pr), which also does the whitespace fixes.

silverwind commented:
Got to correct myself, `apply-pr` by itself does no whitespace fixing, but having

```
git config --global apply.whitespace fix
```

does so, and works for both `apply-pr` and `git am` 😉
